### PR TITLE
Implemented Popcnt function

### PR DESCRIPTION
**Pull request by parth121101**
fixes: #3653

### CODE REVIEW

File: src/libasr/pass/intrinsic_functions.h [hunk 1]
```diff
@@ -2579,6 +2579,94 @@ namespace Mod {
...
+                }
+                return make_ConstantWithType(make_IntegerConstant_t, count, t1, loc);
+            } else {
+                int count = 0;
+                while (val) {
+                    count += val & 1; 
+                    val >>= 1; 
+                }
+                return make_ConstantWithType(make_IntegerConstant_t, count, t1, loc);
+            }
```
Comment (Shaikh-Ubaid) [inline src/libasr/pass/intrinsic_functions.h:2606]: This logic seems duplicated below, so we can make it a function. You can use a templated function to handle integers of different kinds.

File: src/libasr/pass/intrinsic_functions.h [hunk 2]
```diff
@@ -2579,6 +2579,94 @@ namespace Mod {
...
+        }
```
Comment (Shaikh-Ubaid) [inline src/libasr/pass/intrinsic_functions.h:2627]: Also, why not use `__builtin_popcount` and `__builtin_popcountll`?

File: src/libasr/pass/intrinsic_functions.h [hunk 3]
```diff
@@ -2579,6 +2579,94 @@ namespace Mod {
...
+        Vec<ASR::call_arg_t>& new_args, int64_t /*overload_id*/) {
+            declare_basic_variables("_lcompilers_popcnt_" + type_to_str_python(arg_types[0]));
+        fill_func_arg("i", arg_types[0]);
+        auto result = declare(fn_name, return_type, ReturnVar);
+        int32_t kind = extract_kind_from_ttype_t(arg_types[0]);
+
+        auto count = declare("j", ASRUtils::TYPE(ASR::make_Integer_t(al, loc, kind)), Local);
+        auto val = declare("k", ASRUtils::TYPE(ASR::make_Integer_t(al, loc, kind)), Local);
+        auto mask = declare("l", ASRUtils::TYPE(ASR::make_Integer_t(al, loc, kind)), Local);
+        body.push_back(al, b.Assignment(count, b.i(0,arg_types[0])));
```
Comment (Shaikh-Ubaid) [inline src/libasr/pass/intrinsic_functions.h:2642]: Please write in a comment here the definition of the function in `fortran`. That makes it easier to understand the logic implemented in ASR and also verify it.

**Review by Shaikh-Ubaid — COMMENTED** [review]
I shared few comments. Rest looks good to me. Thanks for the implementation and Great work!

**Comment by Shaikh-Ubaid** [comment]
Please mark as "Ready for review" when ready.

**Comment by parth121101** [comment]
> Please mark as "Ready for review" when ready.

@Shaikh-Ubaid I have applied the code review as suggested by you. One of the tests is failing. Can you please let me know what is the reason behind it?

**Comment by Pranavchiku** [comment]
Error: `error C3861: '__builtin_popcountll': identifier not found`, you'll have to include corresponding header.

**Comment by Shaikh-Ubaid** [comment]
>One of the tests is failing. Can you please let me know what is the reason behind it?

You can try the following in that specific order.
1. Try https://stackoverflow.com/a/24550632
2. Try switching to `std::popcnt` as Pranav shared above. Get the linux, mac and windows tests passed. You can see which one fails in the Github CI and debug it.
3. If none of the above works, you can fallback to your custom code (refactored into functions).

Most likely, I believe 1. or 2. would work, but if not 3. can be the possible last option.

**Comment by parth121101** [comment]
> > One of the tests is failing. Can you please let me know what is the reason behind it?
> 
> You can try the following in that specific order.
> 
> 1. Try https://stackoverflow.com/a/24550632
> 2. Try switching to `std::popcnt` as Pranav shared above. Get the linux, mac and windows tests passed. You can see which one fails in the Github CI and debug it.
> 3. If none of the above works, you can fallback to your custom code (refactored into functions).
> 
> Most likely, I believe 1. or 2. would work, but if not 3. can be the possible last option.

@Shaikh-Ubaid both of the approaches doesn't work. I need to move ahead with the scratch implementation.

File: src/libasr/pass/intrinsic_functions.h [hunk 4]
```diff
@@ -2579,6 +2578,99 @@ namespace Mod {
...
+        std::vector<ASR::stmt_t*> while_loop_body_neg;
+        while_loop_body_neg.push_back(b.If(iNotEq(b.i(0,arg_types[0]), (b.i_BitAnd(val,mask, arg_types[0]))), {b.Assignment(count, b.i_tAdd(count, b.i(1, arg_types[0]), arg_types[0]))},
+                                {}));
+        while_loop_body_neg.push_back(b.Assignment(mask, b.i_BitLshift(mask, b.i(1, arg_types[0]), arg_types[0])));
+
+        body.push_back(al, b.If((iGtE(args[0], b.i(0,arg_types[0]))), {b.While(whileloop_test_pos, while_loop_body_pos)},
+                                {b.While(whileloop_test_neg, while_loop_body_neg)}));
+
+        ASR::expr_t *cast = ASRUtils::EXPR(ASR::make_Cast_t(al, loc, count, ASR::cast_kindType::IntegerToInteger, return_type, nullptr));
+        body.push_back(al, b.Assignment(result, cast));
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_functions.h:2666]: Avoid using this kind of cast expr node, use `b.i2i`

Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_functions.h:2666]: Please apply the following diff:

```diff
--- a/src/libasr/pass/intrinsic_functions.h
+++ b/src/libasr/pass/intrinsic_functions.h
@@ -2576,6 +2576,9 @@ namespace Mod {
         return b.Call(f_sym, new_args, return_type, nullptr);
     }
 
+    static inline ASR::expr_t* MOD(ASRBuilder &b, ASR::expr_t* a, ASR::expr_t* p, SymbolTable* scope) {
+        return b.CallIntrinsic(scope, {expr_type(a), expr_type(p)}, {a, p}, expr_type(a), 0, Mod::instantiate_Mod);
+    }
 } // namespace Mod
 
 namespace Popcnt {
@@ -2647,23 +2650,19 @@ namespace Popcnt {
         body.push_back(al, b.Assignment(val, args[0]));
         body.push_back(al, b.Assignment(mask, b.i(1,arg_types[0])));
 
-        ASR::expr_t* whileloop_test_pos = iNotEq(val, b.i(0, arg_types[0]));
-        std::vector<ASR::stmt_t*> while_loop_body_pos;
-        while_loop_body_pos.push_back(b.Assignment(count, b.i_tAdd(count, b.CallIntrinsic(scope, {arg_types[0], arg_types[0]},
-                                    {val, b.i(2,arg_types[0])}, arg_types[0], 0, Mod::instantiate_Mod), arg_types[0])));
-        while_loop_body_pos.push_back(b.Assignment(val, b.i_BitRshift(val, b.i(1, arg_types[0]), arg_types[0])));
-
-        ASR::expr_t* whileloop_test_neg = iNotEq(mask, b.i(0, arg_types[0]));
-        std::vector<ASR::stmt_t*> while_loop_body_neg;
-        while_loop_body_neg.push_back(b.If(iNotEq(b.i(0,arg_types[0]), (b.i_BitAnd(val,mask, arg_types[0]))), {b.Assignment(count, b.i_tAdd(count, b.i(1, arg_types[0]), arg_types[0]))},
-                                {}));
-        while_loop_body_neg.push_back(b.Assignment(mask, b.i_BitLshift(mask, b.i(1, arg_types[0]), arg_types[0])));
-
-        body.push_back(al, b.If((iGtE(args[0], b.i(0,arg_types[0]))), {b.While(whileloop_test_pos, while_loop_body_pos)},
-                                {b.While(whileloop_test_neg, while_loop_body_neg)}));
-
-        ASR::expr_t *cast = ASRUtils::EXPR(ASR::make_Cast_t(al, loc, count, ASR::cast_kindType::IntegerToInteger, return_type, nullptr));
-        body.push_back(al, b.Assignment(result, cast));
+        body.push_back(al, b.If(iGtE(args[0], b.i(0,arg_types[0])), {
+            b.While(iNotEq(val, b.i(0, arg_types[0])), {
+                b.Assignment(count, b.i_tAdd(count, Mod::MOD(b, val, b.i(2, arg_types[0]), scope), arg_types[0])),
+                b.Assignment(val, b.i_tDiv(val, b.i(2, arg_types[0]), arg_types[0]))
+                })
+        }, {
+            b.While(iNotEq(mask, b.i(0, arg_types[0])), {
+                b.If(iNotEq(b.i(0,arg_types[0]), (b.i_BitAnd(val,mask, arg_types[0]))), {b.Assignment(count, b.i_tAdd(count, b.i(1, arg_types[0]), arg_types[0]))},
+                            {}),
+                b.Assignment(mask, b.i_BitLshift(mask, b.i(1, arg_types[0]), arg_types[0]))
+            })
+        }));
+        body.push_back(al, b.Assignment(result, b.i2i(count, return_type)));
         ASR::symbol_t *f_sym = make_ASR_Function_t(fn_name, fn_symtab, dep, args, body, result, ASR::abiType::Source, ASR::deftypeType::Implementation, nullptr);
         scope->add_symbol(fn_name, f_sym);
         return b.Call(f_sym, new_args, return_type, nullptr);
```

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
Requested changes

File: src/libasr/pass/intrinsic_functions.h [hunk 5]
```diff
@@ -2576,9 +2575,101 @@ namespace Mod {
...
+                mask = mask * 2
+            end do
+        end if
+        r = count
+        end function popcnt
+        */
+
+        auto count = declare("j", ASRUtils::TYPE(ASR::make_Integer_t(al, loc, kind)), Local);
+        auto val = declare("k", ASRUtils::TYPE(ASR::make_Integer_t(al, loc, kind)), Local);
+        auto mask = declare("l", ASRUtils::TYPE(ASR::make_Integer_t(al, loc, kind)), Local);
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_functions.h:2648]:
```suggestion
        auto count = declare("j", arg_types[0], Local);
        auto val = declare("k", arg_types[0] Local);
        auto mask = declare("l", arg_types[0], Local);
```

Comment (Pranavchiku) [reply src/libasr/pass/intrinsic_functions.h:2648]: I believe this will work, try locally once.

File: src/libasr/pass/intrinsic_functions.h [hunk 6]
```diff
@@ -2576,9 +2575,101 @@ namespace Mod {
...
+        return make_ConstantWithType(make_IntegerConstant_t, count, t1, loc);
+    }
+
+    static inline ASR::expr_t* instantiate_Popcnt(Allocator &al, const Location &loc,
+        SymbolTable *scope, Vec<ASR::ttype_t*>& arg_types, ASR::ttype_t *return_type,
+        Vec<ASR::call_arg_t>& new_args, int64_t /*overload_id*/) {
+            declare_basic_variables("_lcompilers_popcnt_" + type_to_str_python(arg_types[0]));
+        fill_func_arg("i", arg_types[0]);
+        auto result = declare(fn_name, return_type, ReturnVar);
+        int32_t kind = extract_kind_from_ttype_t(arg_types[0]);
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_functions.h:2620]:
```suggestion
```

If the above given changes work we do not need this,

File: src/libasr/pass/intrinsic_functions.h (hunk 5 above)
Comment (parth121101) [reply src/libasr/pass/intrinsic_functions.h:2648]: It doesnt work

Comment (Pranavchiku) [reply src/libasr/pass/intrinsic_functions.h:2648]: It works fine for me:

```diff
--- a/src/libasr/pass/intrinsic_functions.h
+++ b/src/libasr/pass/intrinsic_functions.h
@@ -2617,7 +2617,6 @@ namespace Popcnt {
             declare_basic_variables("_lcompilers_popcnt_" + type_to_str_python(arg_types[0]));
         fill_func_arg("i", arg_types[0]);
         auto result = declare(fn_name, return_type, ReturnVar);
-        int32_t kind = extract_kind_from_ttype_t(arg_types[0]);
         /*
         function popcnt(i) result(r)
         integer, intent(in) :: i
@@ -2643,9 +2642,9 @@ namespace Popcnt {
         end function popcnt
         */
 
-        auto count = declare("j", ASRUtils::TYPE(ASR::make_Integer_t(al, loc, kind)), Local);
-        auto val = declare("k", ASRUtils::TYPE(ASR::make_Integer_t(al, loc, kind)), Local);
-        auto mask = declare("l", ASRUtils::TYPE(ASR::make_Integer_t(al, loc, kind)), Local);
+        auto count = declare("j", arg_types[0], Local);
+        auto val = declare("k", arg_types[0], Local);
+        auto mask = declare("l", arg_types[0], Local);
         body.push_back(al, b.Assignment(count, b.i(0,arg_types[0])));
         body.push_back(al, b.Assignment(val, args[0]));
         body.push_back(al, b.Assignment(mask, b.i(1,arg_types[0])));
```

```console
% ./integration_tests/run_tests.py
[....]

100% tests passed, 0 tests failed out of 876

Label Time Summary:
c                 =   0.26 sec*proc (72 tests)
c_nopragma        =   0.02 sec*proc (3 tests)
cpp               =   0.15 sec*proc (44 tests)
fortran           =   0.39 sec*proc (97 tests)
gfortran          =   3.51 sec*proc (838 tests)
llvm              =   3.67 sec*proc (876 tests)
llvm2             =   0.04 sec*proc (11 tests)
llvmImplicit      =   0.01 sec*proc (2 tests)
llvmStackArray    =   0.44 sec*proc (110 tests)
llvm_nopragma     =   0.02 sec*proc (3 tests)
wasm              =   0.80 sec*proc (212 tests)
x86               =   0.06 sec*proc (19 tests)

Total Test time (real) =   0.49 sec
```

```console
% ./run_tests.py
[....]
errors/kind_star_of_logical.f90 * asr ✓ 
errors/array_constructor_with_different_kind.f90 * asr ✓ 
errors/incompatible_dimension_assignment_arr1.f90 * asr ✓ 
errors/incompatible_ranks_allocatable_arr2.f90 * asr ✓ 
errors/end_function_name_not_match.f90 * ast ✓ 
errors/incorrect_number_args_subroutine.f90 * asr ✓ 
errors/incorrect_number_args_function.f90 * asr ✓ 
errors/class_procedure_extra_args.f90 * asr ✓ 
errors/more_kwargs_than_acceptable_to_subroutine.f90 * asr ✓ 
optional_argument_subroutine_in_type.f90 * asr ✓ 
errors/runtime_stacktrace_01.f90 * run_dbg ✓ 
TESTS PASSED
```

**Review by certik — APPROVED** [review]
I think that this is good enough to merge. The implementation can be possibly improved, but that can be done later as needed. For example the ASR implementation can also use ">>", just like the compile time implementation.

**Comment by certik** [comment]
I don't know why the test fails, that will have to be fixed.

**Review by Pranavchiku — APPROVED** [review]
Good to merge. Thanks @parth121101 !

**Comment by Pranavchiku** [comment]
@parth121101 do you wish to polish commit history or atm I will prefer to squash and merge?

**Comment by parth121101** [comment]
> @parth121101 do you wish to polish commit history or atm I will prefer to squash and merge?

will do it, give me a moment

**Pull request merged**